File: sakila-film-service/src/main/java/com/example/sakila/event/bus/EventBus.java
package com.example.sakila.event.bus;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class EventBus {

  private final Logger log = LoggerFactory.getLogger(EventBus.class);

  private final String name;

  private final Map<Class<?>, List<HandlerMethod>> handlers;

  public EventBus(String name) {
    this.name = name;
    handlers = new ConcurrentHashMap<>();
  }

  public boolean hasHandlerForType(Class type) {
    return handlers.keySet().contains(type);
  }

  public void emit(Object object) {
    Class<?> type = object.getClass();

    if (handlers.containsKey(type)) {
      handlers.get(type).forEach(h -> {
        log.info("Invoking {}::{} asynchronously", h.getOwner().getClass().getName(), h.getMethod().getName());
        CompletableFuture.runAsync(() -> h.invoke(object));
      });
    }
  }

  public void emitSynchronously(Object object) {
    Class<?> type = object.getClass();

    if (handlers.containsKey(type)) {
      handlers.get(type).forEach(h -> {
        log.info("Invoking {}::{} synchronously", h.getOwner().getClass().getName(), h.getMethod().getName());
        h.invoke(object);
      });
    }
  }

  public void register(Object object) {
    log.info("Registering {}", object.getClass().getName());
    List<HandlerMethod> eventHandlers = getEventHandlerMethods(object);

    eventHandlers.forEach(h -> {
      if (!handlers.containsKey(h.getParameterType())) {
        handlers.put(h.getParameterType(), new ArrayList<>());
      }

      handlers.get(h.getParameterType()).add(h);
      log.info("Registered method {}", h.getMethod().getName());
    });
    log.info("Registered {}", object.getClass().getName());
  }

  private List<HandlerMethod> getEventHandlerMethods(Object object) {
    return Arrays.stream(object.getClass().getDeclaredMethods())
        .filter(m -> m.isAnnotationPresent(Handler.class))
        .map(m -> {
          checkEventHandlerMethodParameters(m);
          return new HandlerMethod(object, m, m.getParameterTypes()[0]);
        })
        .collect(Collectors.toList());
  }

  private void checkEventHandlerMethodParameters(Method method) {
    if (method.getParameterCount() > 1) {
      throw new IllegalArgumentException("Method annotated with @Handler has more than 1 parameter");
    }
  }

  public String getName() {
    return name;
  }
}

File: sakila-payment-service/src/main/java/com/example/sakila/event/bus/EventBus.kt
package com.example.sakila.event.bus

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory
import java.lang.IllegalArgumentException
import java.lang.reflect.Method
import java.util.concurrent.ConcurrentHashMap

class EventBus(private val name: String) {

  private val log = LoggerFactory.getLogger(EventBus::class.java)

  private val handlers: MutableMap<Class<*>, MutableList<HandlerMethod>>

  init {
    this.handlers = ConcurrentHashMap()
  }

  fun hasHandlerForType(type: Class<*>): Boolean = handlers.containsKey(type)

  fun emit(payload: Any) {
    val type = payload::class.java

    if (handlers.containsKey(type)) {
      handlers[type]!!.forEach {
        log.info("Invoking {}::{} asynchronously", it.owner::class.java.name, it.method.name)
        GlobalScope.launch { it.invoke(payload)
        }
      }
    }
  }

  fun emitSynchronously(payload: Any) {
    val type = payload::class.java

    if (handlers.containsKey(type)) {
      handlers[type]!!.forEach {
        log.info("Invoking {}::{} synchronously", it.owner::class.java.name, it.method.name)
        it.invoke(payload)
      }
    }
  }

  private fun checkEventHandlerMethodParameters(method: Method) {
    if (method.parameterCount > 1) {
      throw IllegalArgumentException("Method annotated with @Handler has more than 1 parameter")
    }
  }

  private fun getEventHandlerMethods(component: Any): List<HandlerMethod> {
    return component::class.java.declaredMethods
        .filter { it.isAnnotationPresent(Handler::class.java) }
        .map {
          checkEventHandlerMethodParameters(it)
          HandlerMethod(component, it, it.parameterTypes[0])
        }
  }

  fun register(component: Any) {
    val eventHandlers = getEventHandlerMethods(component)

    eventHandlers.forEach { h ->
      if (handlers.containsKey(h.parameterType)) {
        handlers[h.parameterType]!!.add(h)
      } else {
        handlers[h.parameterType] = mutableListOf(h)
      }
    }
  }
}
